enhancements/network/egress-qos.md

```diff
@@ -0,0 +1,119 @@
+
+By introducing a new CRD `EgressQos`, users could specify a DSCP value for packets originating from pods on a given namespace heading to a specified CIDR.
+The CRD will be Namespaced, with one resource allowed per namespace.
+The resources will be watched by ovn-k, which in turn will configure OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#QoS_TABLE). 
+
+## Motivation
+
+Telco customers require support for DSCP marking capability for some of their 5G applications, giving some pods precedence over others.
+The QoS markings will be consumed and acted upon by objects outside of the OpenShift cluster to optimize traffic flow throughout their networks.
+
+### Goals
+
+- Provide a mechanism for users to set DSCP on egress traffic coming from specific namespaces.
+
+### Non-Goals
+
+- Ingress QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth QoS mechanism.
+
+- The DSCP marking does not need to be handled or acted upon by OpenShift, just added to selected headers.
+
+- Marking East/West traffic, exposing the DSCP value from the inner packet to the outer geneve packet.
+
+## Proposal
+
+To achieve egress DSCP marking on pods, we introduce a new CRD `EgressQos`, which allows specifying a set of DSCP values and destination CIDRs - meaning traffic coming from pods on the namespace heading to each destination CIDR will be marked with the given DSCP value.
+
+```yaml
+kind: EgressQos
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 46
+    dstCIDR: 0.0.0.0/0
+    priority: "?" # Do we add a priority field to handle overlapping CIDRs? Should we handle that in the code instead?
+  - dscp: 30
+    dstCIDR: 1.2.3.4/32
+```
+
+OVN-K watches these resources, creating/updating/deleting QoS objects in nbctl accordingly.
+For example, assuming there's a single pod `app` in namespace `default` on node `node1` and the above `EgressQos` created the equivalent of:
+```bash
+ovn-nbctl qos-add node1 from-lport 1 "ip4.src == <default_ns_address_set> && ip4.dst == 0.0.0.0/0" dscp=46
+ovn-nbctl qos-add node1 from-lport 2 "ip4.src == <default_ns_address_set> && ip4.dst == 1.2.3.4/32" dscp=30
+```
+will be executed.
+
+In addition it'll watch pods to decide if further updates are needed, for example attaching the QoS to another switch.
+
+### User Stories
+
+### API Extensions
+
+A new namespace-scoped CRD is introduced.
+### Implementation Details/Notes/Constraints
+
+- Adding a new API `EgressQos` under the `k8s.ovn.org/v1` version to `pkg/crd`.
+
+- Adding a controller that watches `EgressQoses`, `Pods`, producing the relevant QoS objects in nbdb and attaching them to the join switch or to the relavant logical switches in LGW.
```

Comment:
I know we spoke about this offline, but we need to decide if its going to be on node switch or join switch depending on the gateway or if there isn't much perf impact for both modes we can probably leave it on the node switch.

Reply:
oh @tssurya I was thinking about the same exact thing before I read your comment, since it seems we will move the egress ACL to nodeSwitch then it seems like having QoS there also will make sense ?